FFmpegInterop/H264SampleProvider.cpp

```diff
@@ -18,55 +18,209 @@
+				tie(m_spsPpsData, m_spsPpsNaluLengths) = parser.GetSpsPpsData();
+			}
+		}
+		else
+		{
+			TraceLoggingWrite(g_FFmpegInteropProvider, "EmptyCodecPrivate", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+				TraceLoggingValue(m_stream->index, "StreamId"));
+		}
+	}
+
+	void H264SampleProvider::SetEncodingProperties(_Inout_ const IMediaEncodingProperties& encProp, _In_ bool setFormatUserData)
 	{
-		hr = GetSPSAndPPSBuffer(dataWriter);
+		NALUSampleProvider::SetEncodingProperties(encProp, setFormatUserData);
+
+		if (!m_isBitstreamAnnexB)
+		{
+			AVCConfigParser parser{ m_stream->codecpar->extradata, static_cast<uint32_t>(m_stream->codecpar->extradata_size) };
+			if (parser.HasNoFMOASO())
+			{
+				MediaPropertySet properties{ encProp.Properties() };
+				properties.Insert(MF_MT_VIDEO_H264_NO_FMOASO, PropertyValue::CreateUInt32(true));
+			}
+		}
 	}
 
-	if (SUCCEEDED(hr))
+	AVCConfigParser::AVCConfigParser(_In_reads_(dataSize) const uint8_t* data, _In_ uint32_t dataSize) :
+		m_data(data),
+		m_dataSize(dataSize)
 	{
-		// Call base class method that simply write the packet to stream as is
-		hr = MediaSampleProvider::WriteAVPacketToStream(dataWriter, avPacket);
+		// Validate parameters
+		WINRT_ASSERT(m_data != nullptr);
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_dataSize < MIN_SIZE);
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_data[0] != 1);
 	}
 
-	// We have a complete frame
-	return hr;
-}
+	uint8_t AVCConfigParser::GetNaluLengthSize() const noexcept
+	{
+		return m_data[4] & 0x03 + 1;
+	}
 
-HRESULT H264SampleProvider::GetSPSAndPPSBuffer(DataWriter^ dataWriter)
-{
-	HRESULT hr = S_OK;
+	tuple<vector<uint8_t>, vector<uint32_t>> AVCConfigParser::GetSpsPpsData() const
+	{
+		vector<uint8_t> spsPpsData;
+		vector<uint32_t> spsPpsNaluLengths;
+		uint32_t pos{ 5 };
+
+		// Parse SPS NALUs
+		uint8_t spsCount{ static_cast<uint8_t>(m_data[pos++] & 0x1F) };
+		pos += ParseParameterSets(spsCount, m_data + pos, m_dataSize - pos, spsPpsData, spsPpsNaluLengths);
+
+		// Parse PPS NALUs
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, pos >= m_dataSize);
+		uint8_t ppsCount{ m_data[pos++] };
+		ParseParameterSets(ppsCount, m_data + pos, m_dataSize - pos, spsPpsData, spsPpsNaluLengths);
+
+		return { spsPpsData, spsPpsNaluLengths };
+	}
+
+	uint32_t AVCConfigParser::ParseParameterSets(
+		_In_ uint8_t parameterSetCount,
+		_In_reads_(dataSize) const uint8_t* data,
+		_In_ uint32_t dataSize,
+		_Inout_ vector<uint8_t>& spsPpsData,
+		_Inout_ vector<uint32_t>& spsPpsNaluLengths) const
+	{
+		// Reserve estimated space now to minimize reallocations
+		spsPpsNaluLengths.reserve(spsPpsNaluLengths.size() + parameterSetCount);
+
+		// Parse the parameter sets and convert the SPS/PPS data to Annex B format
+		uint32_t pos{ 0 };
+		for (uint8_t i{ 0 }; i < parameterSetCount; i++)
+		{
+			// Get the NALU length
+			uint32_t naluLength{ GetAVCNaluLength(data + pos, dataSize - pos, sizeof(uint16_t)) };
```

Comment:
>dataSize - pos [](start = 53, length = 14)

Need to ensure that pos doesn't exceed dataSize prior to this call (to prevent an underflow here and a subsequent OOB read in GetAVCNaluLength).

Reply:
On the first iteration pos is 0 so it's guaranteed to be <= dataSize. On subsequent iterations pos is guaranteed to be <= dataSize due to the bounds check in GetAVCNaluLength().

Reply:
Yea, I missed the bounds check in GetAVCNaluLength().